smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/core/KotlinWriter.kt

```diff
@@ -383,6 +296,27 @@ class KotlinPropertyFormatter(
     }
 }
 
+// Specifies a function that receives a [KotlinWriter]
+typealias InlineKotlinWriter = KotlinWriter.() -> Unit
+/**
+ * Formatter to enable passing a writing function
+ * @param codeWriterCreator function that creates a new [CodeWriter] instance used to generate output of inline content
+ */
+class InlineKotlinWriterFormatter(private val parent: KotlinWriter) : BiFunction<Any, String, String> {
```

Comment:
Why do we need two different versions of this? I think we ought to maybe discourage the use of a raw `CodeWriter` and opt for `KotlinWriter` in our codebase.

Reply:
I see we (I) used a raw `CodeWriter` in the gradle codegen. I don't recall why that was done at this point. Might be OBE though 🤷

Reply:
Well `KotlinWriter` does some kotlin-y things (comment, package, imports) that wouldn't make sense in a Gradle file.  As discussed I investigated another approach where we would have a similar type, `GradleWriter` however found  that to be relatively complex for the value and found this approach to be better.

We may at some point wish to have a "CodeWriterPlus" type that provides some of our features on top of the Smithy Java `CodeWriter`, but without tying it specifically to Kotlin language files.  But,  I didn't see particular merit in adding it in this PR.

Reply:
Regarding why we need two types (for `CodeWriter` and `KotlinWriter`), it comes down to the limitations of `typealias` and generics.  There may be a way of consolidating this once we're able to access the settings of `CodeWriter`.